src/rust/src/tracing_ffi.rs

```diff
@@ -0,0 +1,563 @@
+    let handle = unsafe { &mut *handle };
+    let new_filter = unsafe { CStr::from_ptr(new_filter) }
+        .to_str()
+        .expect("new filter should be a valid string");
+
+    let new_filter = EnvFilter::new(new_filter);
+    handle.reload_handle.reload(new_filter).unwrap();
+}
+
+pub struct FfiCallsite {
+    interest: AtomicUsize,
+    meta: Option<Metadata<'static>>,
+    registration: Once,
+    fields: Vec<&'static str>,
+}
+
+impl FfiCallsite {
+    fn new(fields: Vec<&'static str>) -> Self {
+        FfiCallsite {
+            interest: AtomicUsize::new(0),
+            meta: None,
+            registration: Once::new(),
+            fields,
+        }
+    }
+
+    #[inline(always)]
+    fn is_enabled(&self) -> bool {
+        let interest = self.interest();
+        if interest.is_always() {
+            return true;
+        }
+        if interest.is_never() {
+            return false;
+        }
+
+        tracing::dispatcher::get_default(|current| current.enabled(self.metadata()))
+    }
+
+    #[inline(always)]
+    pub fn register(&'static self) {
+        self.registration
+            .call_once(|| tracing::callsite::register(self));
+    }
+
+    #[inline(always)]
+    fn interest(&self) -> Interest {
+        match self.interest.load(Ordering::Relaxed) {
+            0 => Interest::never(),
+            2 => Interest::always(),
+            _ => Interest::sometimes(),
+        }
+    }
+}
+
+impl Callsite for FfiCallsite {
+    fn set_interest(&self, interest: Interest) {
+        let interest = match () {
+            _ if interest.is_never() => 0,
+            _ if interest.is_always() => 2,
+            _ => 1,
+        };
+        self.interest.store(interest, Ordering::SeqCst);
+    }
+
+    #[inline(always)]
+    fn metadata(&self) -> &Metadata<'static> {
+        self.meta.as_ref().unwrap()
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn tracing_callsite(
+    name: *const c_char,
+    target: *const c_char,
+    level: *const c_char,
+    file: *const c_char,
+    line: u32,
+    fields: *const *const c_char,
+    fields_len: usize,
+    is_span: bool,
+) -> *mut FfiCallsite {
+    let name = unsafe { CStr::from_ptr(name) }.to_str().unwrap();
+    let target = unsafe { CStr::from_ptr(target) }.to_str().unwrap();
+    let level = unsafe { CStr::from_ptr(level) }.to_str().unwrap();
+    let file = unsafe { CStr::from_ptr(file) }.to_str().unwrap();
+
+    let fields = unsafe { slice::from_raw_parts(fields, fields_len) };
+    let fields = fields
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_str())
+        .collect::<Result<Vec<_>, _>>()
+        .unwrap();
+
+    let level = level.parse().unwrap();
+
+    // We immediately convert the new callsite into a pointer to erase lifetime
+    // information. The caller MUST ensure that the callsite is stored statically.
+    let site = Box::into_raw(Box::new(FfiCallsite::new(fields)));
+    let site_ref = unsafe { &*site };
```

Comment:
I believe it _may_ be possible to do this slightly more safely with [`Box::leak`](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.leak), although I don't think this is terribly important (the current implementation is fine, it just involves more instances of the `unsafe` word).